README.md
## TileLite

A lightweight, multi-process, tile-server written as a WSGI app.

A 'gateway' tool to [TileStache](http://tilestache.org/) or [TileCache](http://tilecache.org).

### Requires
 * Python
 * Mapnik (>= 2.x) and a Mapnik xml or mml mapfile you wish to serve.

### Installing

    git clone https://github.com/springmeyer/TileLite.git
    cd TileLite
    sudo python setup.py install

### Usage
    
    liteserv.py <xml> [options]

 * Can be deployed with Mod_wsgi using one thread and many processes.
 * Able to read from an optional configuration file for customization of various rendering and caching parameters.

## Configuration

See the notes in the 'docs' folder and the sample 'tilelite.cfg' in the 'utils' folder.

CHANGELOG.txt
TileLite Changelog
------------------

Version 0.1.5, Novemeber 29th 2010:
-----------------------------------

Fix bug in spherical mercator handling

Enable ?x=&y=&z= style requests


Version 0.1.4, Novemeber 29th 2010:
-----------------------------------
  
  * Support for Mapnik2
  
  * Support for Cascadenik 1.x
  
  * Ensured the liteserv.py development server runs in single thread
  
  * Better detection if map srs is spherical mercator to avoid setting it if not needed
  
  * Faster blank tile response, using rendering no-op
  

Version 0.1.3, September 24th 2009:
---------------------------------
  
  * Improved response handling and demo application

  * Improved support for Mapnik 0.5.1


Version 0.1.2, August 30th 2009:
--------------------------------

  * Make extra data available to installer


Version 0.1.1, August 30th 2009:
--------------------------------

  * Added a 'demo' directory with sample shp, mapnik xml, and html

  * Enabled support for Mapnik 0.5.1 (Previously only worked with >= 0.6.0)


Version 0.1.0, May 3rd 2009:
----------------------------

 * Tagged first release.

deploy.py

#!/usr/bin/env python

"""
Release Steps
-------------
 * write 'setup.py' file
 * create ~./pypirc file with pypi user:pass
 * register package with: 'python setup.py register'
 * Edit CHANGELOG.txt
 * Increment '__version__' in main script and 'version' in setup.py
 * Run `deploy.py` to create sdist, upload, and create tag
 * Commit tag
 * Update wiki
"""

import sys
import time
from subprocess import call as subcall

app = 'tilelite'
version = __import__(app).__version__

def call(cmd):
  try:
    response = subcall(cmd,shell=True)
    print
    time.sleep(1)
    if response < 0:
      sys.exit(response)
  except OSError, E:
    sys.exit(E)

def cleanup():
    call('sudo rm -rf *.egg* *.pyc dist/ build/')

def tag():
    call('hg tag -u springmeyer -m "tagging the %s release" %s ' % (version,version))

def upload():
    call('python setup.py sdist upload')

def main():
    cleanup()
    tag()
    upload()
    cleanup()
    
if __name__ == '__main__':
    main()

liteserv.py
#!/usr/bin/env python

import os
import sys
import socket
import logging
from optparse import OptionParser
from wsgiref.simple_server import make_server, WSGIServer, WSGIRequestHandler

CONFIG = 'tilelite.cfg'
MAP_FROM_ENV = 'MAPNIK_MAP_FILE'
    
parser = OptionParser(usage="""
    python liteserv.py <mapfile.xml> [options]
    """)

parser.add_option('-i', '--ip', default='0.0.0.0', dest='host',
    help='Specify a ip to listen on (defaults to 0.0.0.0/localhost)'
    )

parser.add_option('-p', '--port', default=8000, dest='port', type='int',
    help='Specify a custom port to run on: eg. 8080'
    )

parser.add_option('--config', default=None, dest='config',
    help='''Specify the use of a custom TileLite config file to override default settings. By default looks for a file locally called 'tilelite.cfg'.'''
    )

parser.add_option('-s', '--size', default=None, dest='size', type='int',
    help='Specify a custom tile size (defaults to 256)'
    )

parser.add_option('-b', '--buffer-size', default=None, dest='buffer_size', type='int',
    help='Specify a custom map buffer_size (defaults to 128)'
    )

parser.add_option('-z', '--max-zoom', default=None, dest='max_zoom', type='int',
    help='Max zoom level to support (defaults to 22)'
    )
    
parser.add_option('-f', '--format', default=None, dest='format',
    help='Specify a custom image format (png or jpeg) (defaults to png)'
    )

parser.add_option('--paletted', default=False, dest='paletted', action='store_true',
    help='Use paletted/8bit PNG (defaults to False)'
    )

parser.add_option('-d','--debug', default=True, dest='debug', type="choice", choices=('True','False'),
    help='Run in debug mode (defaults to True)'
    )

parser.add_option('-c','--caching', default=False, dest='caching', action='store_true',
    help='Turn on tile caching mode (defaults to False)'
    )

parser.add_option('--cache-path', default=None, dest='cache_path',
    help='Path to tile cache directory (defaults to "/tmp")'
    )

parser.add_option('--cache-force', default=False, dest='cache_force', action='store_true',
    help='Force regeneration of tiles while in caching mode (defaults to False)'
    )

parser.add_option('--processes', default=1, dest='num_processes', type='int',
    help='If werkzeug is installed, number of rendering processes to allow'
    )

parser.add_option('--threaded', default=False, dest='threaded', action='store_true',
    help='If werkzeug is installed, place each request on a new thread'
    )
    
def run(process):
    try:
        process.serve_forever()
    except KeyboardInterrupt:
        process.server_close()
        sys.exit(0)

def strip_opts(options):
    remove = [None,'config','port','host']
    params = {}
    for k,v in options.items():
        if not k in remove and not v is None:
            params[k] = v
    return params

def print_url(options):
    if not application.debug:
        sys.stderr.write('TileLite debug mode is *off*...\n')
    sys.stderr.write("Listening on %s:%s...\n" % (options.host,options.port))
    sys.stderr.write("To access locally view: http://localhost:%s\n" % options.port)
    remote = "To access remotely view: http://%s" % socket.getfqdn()
    if not options.port == 80:
        remote += ":%s" % options.port
    try:
        remote += "\nor view: http://%s" % socket.gethostbyname(socket.gethostname())
    except socket.gaierror:
        pass
    if not options.port == 80:
        remote += ":%s" % options.port
    sys.stderr.write('%s\n' % remote)

if __name__ == '__main__':
    (options, args) = parser.parse_args()
    log = logging.getLogger('tilelite.liteserv')
        
    if len(args) < 1:
        try:
            mapfile = os.environ[MAP_FROM_ENV]
        except:
            sys.exit("\nPlease provide either the path to a mapnik xml or\nset an environment setting called '%s'\n" % (MAP_FROM_ENV))
    else:
        mapfile = args[0]
        if not os.path.exists(mapfile):
            sys.exit('Could not locate mapfile.')
    
    logging.basicConfig(level=(logging.DEBUG if options.debug else logging.INFO))
    log.debug("Using mapfile: '%s'", os.path.abspath(mapfile))
        
    if options.config:
        if not os.path.isfile(options.config):
            sys.exit('That does not appear to be a valid config file')
        else:
            CONFIG = options.config

    if not os.path.exists(CONFIG):
        if options.config:
            sys.exit('Could not locate custom config file')
        else:
            CONFIG = None
    
    if CONFIG:
        log.debug("Using config file: '%s'", os.path.abspath(CONFIG))

    if options.cache_path and not options.caching:
        options.caching = True

    if options.cache_force and not options.caching:
        options.caching = True

    #parser.error("Caching must be turned on with '--caching' flag for liteserv.py to accept '--cache-path' option")
    #http_setup = options.host, options.port
    #httpd = simple_server.WSGIServer(http_setup, WSGIRequestHandler)
    #httpd.set_app(application)

    from tilelite import Server
    application = Server(mapfile, CONFIG)
    application.absorb_options(strip_opts(options.__dict__))
                
    try:
        from werkzeug import run_simple
        print_url(options)
        run_simple(options.host, options.port, application, threaded=options.threaded, processes=options.num_processes)
    except:
        if options.num_processes > 1:
            sys.exit('The werkzeug python server must be installed to run multi-process\n')
        sys.stderr.write('Note: werkzeug is not installed so falling back to built-in python wsgiref server.\n')
        sys.stderr.write('Install werkzeug from http://werkzeug.pocoo.org/\n\n')
        
        from wsgiref import simple_server
        # below code was for testing multi-threaded rendering
        # which only works if we copy a map object per thread
        # so avoid this and only run multiprocess...
        #from SocketServer import ThreadingMixIn
        #class myServer(ThreadingMixIn, simple_server.WSGIServer):
        #    pass 
        #httpd = myServer(('',options.port), simple_server.WSGIRequestHandler,)
        #httpd.set_app(application)
        httpd = make_server(options.host, options.port, application)        
        print_url(options)
        run(httpd)
        


tiles_app.py
"""
To setup TileLite on a production server using Apache and ModWSGI
create a virtualhost or otherwise insert the WSGI configuration into
your Apache configuration like so:

    WSGIScriptAlias /<url> /path/to/this/tiles_app.py
    WSGIDaemonProcess <process name> user=<user> group=<group> processes=10 threads=1
    WSGIProcessGroup <process name>

 * 'tiles_app.py' is the name of the simple python script below that associates the 
   tilelite.Server instance with a Mapnik xml file. It can be named anything you like
   but should end with either a '.wsgi' or '.py' extension.
    
 * <url> can be either be '/' (to mount the script at http://yourserver.com/) or it can be 
   a path such as '/tiles' to mount the server at http://yourserver.com/tiles

 * <process name> can be any unique name like 'tileliteserver'

 * <user> and <group> should be a unix user that has permissions to the 'tiles_app.py'

 * Note: this is a multiprocess (not threaded) server so you *can* set 'processes' >= 1
   but threads *must be* == 1, otherwise this server will not work within Apache.

An example setup would be:

    ## TileLite sample setup ##
    WSGIScriptAlias /tiles /home/mapnik/projects/tilelite/tiles_app.py
    WSGIDaemonProcess tileliteserver user=www-data group=www-data processes=10 threads=1
    WSGIProcessGroup tileliteserver

Next, edit the script code below and place it where the WSGIScriptAlias path points to.

Then test your apache configuration and restart. On debian linux this might look like:

    $ sudo apache2ctl configtest
    $ /etc/init.d/apache restart

Then go to:

    http://yourserver.com/tiles/
    
"""

def append_local_dir():
    import os,sys
    local_dir = os.path.dirname(__file__)
    sys.path.append(local_dir)

# put local tilelite on PYTHONPATH
# only needed if you have not already installed in site-packages
#append_local_dir()

from tilelite import Server

# point at your mapnik xml or cascadenik mml...
mapfile = '/home/mapnik/osm.xml'
config = None
# note: this variable must be called 'application'
application = Server(mapfile,config)


tilelite.py
#!/usr/bin/env python

__author__ = 'Dane Springmeyer (dbsgeo [ -a- ] gmail.com)'
__copyright__ = 'Copyright 2009, Dane Springmeyer'
__version__ = '0.1.5'
__license__ = 'BSD'

import os
#import re
import sys
import time
import math
import urllib
import tempfile
import logging

try:
    import mapnik2 as mapnik
except ImportError:
    import mapnik

# http://spatialreference.org/ref/epsg/3785/proj4/
#"+proj=merc +lon_0=0 +k=1 +x_0=0 +y_0=0 +a=6378137 +b=6378137 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"
# QGIS 3785
#"+proj=merc +lon_0=0 +k=1 +x_0=0 +y_0=0 +a=6378137 +b=6378137 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"
# QGIS 4055
#"+proj=longlat +a=6378137 +b=6378137 +towgs84=0,0,0,0,0,0,0 +no_defs"
# proj's WGS 84 / Pseudo-Mercator <3857>
# +proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs <>

# http://spatialreference.org/ref/sr-org/6/
MERC_PROJ4 = "+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +no_defs +over"
MERC_ELEMENTS = dict([p.split('=') for p in MERC_PROJ4.split() if '=' in p])
mercator = mapnik.Projection(MERC_PROJ4)

CSS_STYLE = "font-family: 'Lucida Grande', Verdana, Helvetica, sans-serif; border-left-width: 0px; border-bottom-width: 2px; border-right-width: 0px; border-top-width: 2px; width: 95%; border-color: #a2d545; border-style: solid; font-size: 14px; margin: 10px; padding: 10px; background: #eeeeee; -moz-border-radius: 2px; -webkit-border-radius: 2px;"

#pattern = r'/(?P<version>\d{1,2}\.\d{1,3})\.\d{1,3})/(?P<layername>[a-z]{1,64})/(?P<z>\d{1,10})/(?P<x>\d{1,10})/(?P<y>\d{1,10})\.(?P<extension>(?:png|jpg|gif))'
#request_re = re.compile(pattern)

def is_merc(srs):
    srs = srs.lower()
    if 'epsg:900913' in srs:
        return True
    elif 'epsg:3857' in srs:
        return True
    elif not 'merc' in srs:
        print 'no merc', srs
        return False
    # strip optional modifiers (those without =)
    elements = dict([p.split('=') for p in srs.split() if '=' in p])
    for p in elements:
        if MERC_ELEMENTS.get(p, None) != elements.get(p, None):
            return False
    return True

def parse_config(cfg_file):
    from ConfigParser import SafeConfigParser
    config = SafeConfigParser()
    config.read(cfg_file)
    params = {}
    for section in config.sections():
        options = config.items(section)
        for param in options:
            params[param[0]] = param[1]
    return params

def parse_query(qs):
    query = dict([i.split('=') for i in urllib.unquote(qs).split('&')])
    return query

def is_image_request(path_info):
    if path_info.endswith('.png') | path_info.endswith('.jpeg'):
        return True
    return False

def match(attr,value):
    if isinstance(attr,bool) and str(value).lower() in ['on','yes','y','true']:
        return True
    elif isinstance(attr,bool):
        return False
    elif isinstance(attr,int):
        return int(value)
    elif isinstance(attr,str):
        return value
    else:
        return None
        
class SphericalMercator(object):
    """
    Python class defining Spherical Mercator Projection.
    
    Originally from:  
      http://svn.openstreetmap.org/applications/rendering/mapnik/generate_tiles.py
    """
    def __init__(self,levels=18,size=256):
        self.Bc = []
        self.Cc = []
        self.zc = []
        self.Ac = []
        self.levels = levels
        self.DEG_TO_RAD = math.pi/180
        self.RAD_TO_DEG = 180/math.pi
        self.size = size
        for d in range(0,levels):
            e = size/2.0;
            self.Bc.append(size/360.0)
            self.Cc.append(size/(2.0 * math.pi))
            self.zc.append((e,e))
            self.Ac.append(size)
            size *= 2.0

    @classmethod
    def minmax(a,b,c):
        a = max(a,b)
        a = min(a,c)
        return a

    def ll_to_px(self,px,zoom):
        d = self.zc[zoom]
        e = round(d[0] + px[0] * self.Bc[zoom])
        f = self.minmax(math.sin(DEG_TO_RAD * px[1]),-0.9999,0.9999)
        g = round(d[1] + 0.5 * math.log((1+f)/(1-f))*-self.Cc[zoom])
        return (e,g)
    
    def px_to_ll(self,px,zoom):
        """ Convert pixel postion to LatLong (EPSG:4326) """
        # TODO - more graceful handling of indexing error
        e = self.zc[zoom]
        f = (px[0] - e[0])/self.Bc[zoom]
        g = (px[1] - e[1])/-self.Cc[zoom]
        h = self.RAD_TO_DEG * ( 2 * math.atan(math.exp(g)) - 0.5 * math.pi)
        return (f,h)
    
    def xyz_to_envelope(self,x,y,zoom, tms_style=False, count=1):
        """ Convert XYZ to mapnik.Envelope """
        # flip y to match TMS spec
        if tms_style:
            y = (2**zoom-1) - y
        ll = (x * self.size,(y + count) * self.size)
        ur = ((x + count) * self.size, y * self.size)
        minx,miny = self.px_to_ll(ll,zoom)
        maxx,maxy = self.px_to_ll(ur,zoom)
        lonlat_bbox = mapnik.Envelope(minx,miny,maxx,maxy)
        env = mercator.forward(lonlat_bbox)
        return env

class Server(object):
    def __init__(self, mapfile, config=None):
        # private
        self._changed = []
        self._config = config
        self._locked = False

        self._log = logging.getLogger('tilelite')
        
        # mutable
        self.size = 256
        self.buffer_size = 128
        self.format = 'png'
        self.paletted = False
        self.max_zoom = 22
        self.debug = True
        self.watch_mapfile = False
        self.watch_interval = 2
        self.max_failures = 6

        self.caching = False
        self.cache_force = False
        self.cache_path = '/tmp' #tempfile.gettempdir()
        self.metatile = 0

        self._mapnik_map = mapnik.Map(self.size,self.size)

        if self._config:
            self.absorb_options(parse_config(self._config))

        self._mapfile = mapfile
        if mapfile.endswith('.xml'):
            mapnik.load_map(self._mapnik_map, self._mapfile)
        elif mapfile.endswith('.mml'):
            import cascadenik
            if hasattr(cascadenik,'VERSION'):
                major = int(cascadenik.VERSION.split('.')[0])
                if major < 1:
                    from cascadenik import compile as _compile
                    compiled = '%s_compiled.xml' % os.path.splitext(str(mapfile))[0]
                    open(compiled, 'w').write(_compile(self._mapfile))
                    mapnik.load_map(self._mapnik_map, compiled)
                elif major == 1:
                    if str(mapfile).startswith('http'):
                        output_dir = os.getcwd() #os.path.expanduser('~/.cascadenik')
                    else:
                        output_dir = os.path.dirname(str(mapfile))
                    cascadenik.load_map(self._mapnik_map,mapfile,output_dir,verbose=self.debug)
                elif major > 1:
                    raise NotImplementedError('This TileLite version does not yet support Cascadenik > 1.x, please upgrade to the latest release')
            else:
                from cascadenik import compile as _compile
                compiled = '%s_compiled.xml' % os.path.splitext(str(mapfile))[0]
                open(compiled, 'w').write(_compile(self._mapfile))
                mapnik.load_map(self._mapnik_map, compiled)

        self.post_init_setup()
        
        if self.watch_mapfile:
            self.modified = os.path.getmtime(self._mapfile)
            import thread
            thread.start_new_thread(self.watcher, ())
        self._mapnik_map.zoom_all()
        self.envelope = self._mapnik_map.envelope()

        self.empty_tile = mapnik.Image(self.size, self.size)
        if self._mapnik_map.background:
            self.empty_tile.background = self._mapnik_map.background

    def post_init_setup(self):
        self._merc = SphericalMercator(levels=self.max_zoom+1,size=self.size)
        if not is_merc(self._mapnik_map.srs):
            self._mapnik_map.srs = MERC_PROJ4
            self.msg('Map is not in spherical mercator, so setting that projection....')
                       
    def watcher(self):
        failed = 0
        while 1:
            if not self.modified == os.path.getmtime(self._mapfile):
                self._locked = True
                time.sleep(self.watch_interval/2.0)
                self._log.info('Mapfile **changed**, reloading... ')
                try:
                    self._mapnik_map = mapnik.Map(self.size,self.size)
                    if self._mapfile.endswith('.xml'):
                        mapnik.load_map(self._mapnik_map, self._mapfile)
                    elif self._mapfile.endswith('.mml'):
                        from cascadenik import load_map as load_mml
                        load_mml(self._mapnik_map, self._mapfile)
                    self.msg('Mapfile successfully reloaded from %s' % self._mapfile)
                    if not is_merc(self._mapnik_map.srs):
                        self._mapnik_map.srs = MERC_PROJ4
                        self.msg('Map is not in spherical mercator, so setting that projection....')
                    failed = 0
                except Exception, E:
                    failed += 1
                    again = self.watch_interval*2
                    self.msg('Failed to reload mapfile, will try again in %s seconds:\n%s' % (again,E))
                    time.sleep(again)
                self.modified = os.path.getmtime(self._mapfile)
                self._locked = False
            else:
                time.sleep(self.watch_interval)
            if failed > self.max_failures:
                self.msg('Giving up on mapfile change watching thread...')
                break
        return
    
    def msg(self,message):
        """ WSGI apps must not print to stdout.
        """
        self._log.debug(message)
        #if self.debug:
        #    print >> sys.stderr, '[TileLite Debug] --> %s' % message

    def settings(self):
        settings = '\n'
        for k,v in self.__dict__.items():
            if not k.startswith('_'):
                if k in self._changed:
                    v = '%s *changed' % v
                settings += '%s = %s\n' % (k,v)
        return settings

    def settings_dict(self):
        settings = {}
        for k,v in self.__dict__.items():
            if not k.startswith('_'):
                settings[k] = v
        return settings

    def instance_dict(self):
        d = {}
        m = self._mapnik_map
        m.zoom_all()
        e = m.envelope()
        c = e.center()
        e2 = mercator.inverse(e)
        c2 = e2.center()
        d['extent'] = [e.minx,e.miny,e.maxx,e.maxy]
        d['center'] = [c.x,c.y]
        d['lonlat_extent'] = [e2.minx,e2.miny,e2.maxx,e2.maxy]
        d['lonlat_center'] = [c2.x,c2.y]        
        d['layers'] = [l.name for l in m.layers]
        d['mapfile'] = self._mapfile
        return d
        
    def absorb_options(self,opts):
        """
        """
        for opt in opts.items():
            attr = opt[0]
            if hasattr(self,attr) and not attr.startswith('_'):
                cur = getattr(self,attr)
                new = match(cur,opt[1])
                if not new == cur:
                    setattr(self,attr,new)
                    self._changed.append(attr)
        self.msg(self.settings())

    def ready_cache(self,path_to_check):
        """
        """
        dirname = os.path.dirname(path_to_check)
        if not os.path.exists(dirname):
            os.makedirs(dirname)

    def hit(self,env):
        return self.envelope.intersects(env)

    def cache_save(self, tile_dir, im, fmt):
        self.ready_cache(tile_dir)
        self._log.debug('saving %s: %s', fmt, tile_dir)
        im.save(tile_dir, fmt)

    def cache_fn(self, zoom, x, y, fmt):
        return os.path.join(self.cache_path,str(zoom),str(x),'%s.%s' % (str(y),fmt) )

    def render(self, zoom, x, y, fmt):
        # TODO - throw an error if zoom > self._merc.levels

        envelope = self._merc.xyz_to_envelope(x,y,zoom)
        if not self.hit(envelope): # skip rendering and ds query if tile will be blank
            return self.empty_tile.tostring(fmt)

        tile_dir = self.cache_fn(zoom, x, y, fmt)
        if self.caching and not self.cache_force and os.path.exists(tile_dir):
            # todo: benchmark opening without using mapnik...
            self._log.debug('cache hit: %s', tile_dir)
            im = mapnik.Image(self.size,self.size)
            return im.open(tile_dir).tostring(fmt)

	if self.metatile:
            im = self._render_meta(zoom, x, y, envelope, fmt)
	else:
            im = self._render(zoom, x, y, envelope, fmt)

        if self.caching: self.cache_save(tile_dir, im, fmt)
        return im.tostring(fmt)

    def _render(self, zoom, x, y, envelope, fmt):
        im = mapnik.Image(self.size,self.size)
        self._mapnik_map.zoom_to_box(envelope)
        self._mapnik_map.buffer_size = self.buffer_size
        mapnik.render(self._mapnik_map,im)
        return im

    def _render_meta(self, zoom, x, y, e, fmt):
        s = self.size
        ms = self.metatile
        mx = x - (x % ms) - 1
        my = y - (y % ms) - 1
        self._log.debug("Metatile: %d,%d -> %d,%d", x, y, mx, my)
        envelope = self._merc.xyz_to_envelope(mx,my,zoom, count=ms+2)
        im = mapnik.Image(s * (ms + 2), s * (ms + 2))
        self._mapnik_map.resize(s * (ms + 2), s * (ms + 2))
        self._mapnik_map.zoom_to_box(envelope)
        self._mapnik_map.buffer_size = self.buffer_size
        mapnik.render(self._mapnik_map,im)
        self._hold = im
        self._log.debug("Crop: %d,%d", x - mx, y - my)
        for i in range(1, ms+1):
            for j in range(1, ms+1):
                self.cache_save(self.cache_fn(zoom, mx + i, my + j, fmt),
                                im.view(s * i, s * j, s, s), fmt)
        return im.view(s * (x - mx), s * (y - my), s, s)

    def __call__(self, environ, start_response):
        """ WSGI request handler """
        response_status = "200 OK"
        mime_type = 'text/html'
        fmt = 'png'
        if not self._locked:
            path_info = environ['PATH_INFO']
            qs = environ['QUERY_STRING']    
            zoom,x,y = None,None,None
            if qs:
                query = parse_query(qs)
                x = query.get('x')
                y = query.get('y')
                zoom = query.get('z')
                if x and y and zoom:
                    x = int(x)
                    y = int(y)
                    zoom = int(zoom)
                fmt = query.get('format','png')
            elif is_image_request(path_info):
                uri, fmt = path_info.split('.')
                zoom,x,y = map(int,uri.split('/')[-3:])
            if self.paletted:
                fmt = 'png256'
            if (x is not None) and (y is not None) and (zoom is not None):
                response = self.render(zoom, x, y, fmt)
                mime_type = 'image/%s' % fmt
                self.msg('Zoom,X,Y: %s,%s,%s' % (zoom,x,y,))
                self.msg('scale_denom: %s' % (self._mapnik_map.scale_denominator()))
            elif path_info.endswith('settings/'):
                response = '<h2>TileLite Settings</h2>'
                response += '<pre style="%s">%s</pre>' % (CSS_STYLE,self.settings())
                if self._config:
                    response += '<h4>From: %s</h4>' % self._config
                else:
                    response += '<h4>*default settings*</h4>'
            elif path_info.endswith('settings.json'):
                response = str(self.settings_dict())
                mime_type = 'text/plain'        
            elif path_info.endswith('instance.json'):
                if 'jsoncallback' in query:
                    response = ('%s(%s);' % (query['jsoncallback'], self.instance_dict()))
                    mime_type = 'application/json'
                else:
                    response = str(self.instance_dict())
                    mime_type = 'text/plain'
            elif not path_info.strip('/'):
                root = '%s%s' % (environ['SCRIPT_NAME'], path_info.strip('/'))
                response = '''<h2>TileLite</h2>
                <div style="%(style)s"><p>Welcome, ready to accept a tile request in the format of:</p>
                <p>%(root)s/zoom/x/y.png</p>
                <p>or</p>
                <p>%(root)s/?x=0&y=0&z=0&format=png</p>
                <p>Example url: <a href="%(root)s/1/0/0.png">%(root)s/1/0/0.png</a></p>
                <p>js: var tiles = new OpenLayers.Layer.OSM("Mapnik", "http://%(http_host)s/${z}/${x}/${y}.png");</p>
                <p>See TileLite settings: <a href="%(root)s/settings/">%(root)s/settings/</a>
                | <a href="%(root)s/settings.json">%(root)s/settings.json</a></p>
                <p> More info: <a href="http://bitbucket.org/springmeyer/tilelite/">
                bitbucket.org/springmeyer/tilelite/</a></p></div>
                ''' % {'root': root,'style':CSS_STYLE,'http_host':environ['HTTP_HOST']}
            else:
                response = '<h1> Page Not found </h1>'
                response_status = "404 Not Found"
        else:
            mime_type = 'image/%s' % fmt
            im = mapnik.Image(self.size,self.size)
            im.background = mapnik.Color('pink')
            response = im.tostring(fmt)
                        
        #self.msg('Multithreaded: %s | Multiprocess: %s' % (environ['wsgi.multithread'],environ['wsgi.multiprocess']))
        start_response(response_status,[('Content-Type', mime_type)])
        yield response

if __name__ == '__main__':
    import doctest
    doctest.testmod()
